Add z-index stacking order to vDialog, menuable and associated overlays

Webifi opened the pull request:
Sets z-index on dialogs, associated overlays and menuable elements allowing multiple open dialogs to be stacked based on when the dialog was last set active, not when it was added to the dom.

Also stops propagation of click event on overlay if overlay is used by dialog, stopping the dismissal of all open dialogs if the overlay for one dialog is clicked.

Removes need for #1918 
Fixes #1901 
Fixes #2009 
Fixes #2010 
Fixes #2028

Here's an example of using multiple dialogs and menus **with this PR** (using a build of 0.16.0 with this patch applied):
https://codepen.io/webifi/pen/GMMZOd

And, for comparison, here's using the same markup as above, but **without this PR**:
https://codepen.io/webifi/pen/jGaKjo

Webifi commented:
One thought is how this interacts with v-menu, tooltips, etc.   I haven't looked closely, but I'm guessing the z-index on those are relatively static?  If so, there will be some issues with how these interact.   

Shouldn't be too much of an issue to get them to play nicely with each other.  If you're okay with the method I'm using here (just grabbing the current z-index for active elements directly from the DOM rather than keeping a globally accessible abstracted index of them), I'll put some time in to see what it would take to apply to other applicable components.

jacekkarczmarczyk commented on src/components/VDataTable/__snapshots__/VDataTable.spec.js.snap:
```diff
@@ -87,7 +87,7 @@ exports[`VDataTable.vue should match a snapshot 1`] = `
                   >
                     <div class="menu__content menu__content--select menu__content--auto"
-                         style="max-height: 200px; min-width: 75px; max-width: auto; top: 12px; left: 12px; z-index: 6; display: none;"
+                         style="max-height: 200px; min-width: 75px; max-width: auto; top: 12px; left: 12px; z-index: 0; display: none;"
```
with `z-index: 0` - won't it hide behind the app toolbar for example?

Webifi replied:
It pulls zero when not active.  Once active, it'll z-index in the higher.   I've committed a few more changes.

Webifi commented:
Updated to make menuable and tooltips use a similar dynamic z-index.

Also changed click-outside directive to use capture instead of bubble click events to avoid the click outside being prevented when clicking outside hits an event handler that uses stopPropagation().

johnleider commented on src/components/VDialog/VDialog.js:
```diff
@@ -3,13 +3,22 @@ require('../../stylus/components/_dialogs.styl')
 
 import ClickOutside from '../../directives/click-outside'
 
+import { getZIndex } from '../../util/helpers'
```
Update this to match this spec https://docs.google.com/document/d/1xWWnAw2zYce3sVvdiRC0OuIhNx1PCt-oCNzTXeREFaw/edit?usp=sharing

johnleider commented on src/components/VDialog/VDialog.js:
```diff
@@ -79,9 +95,14 @@ export default {
       ref: 'dialog',
       directives: [
-        { name: 'click-outside', value: this.closeConditional },
+        { name: 'click-outside', value: this.closeConditional, include: () => this.getOpenDependentElements() },
```
break this into new lines

johnleider commented on src/components/VDialog/VDialog.js:
```diff
@@ -79,9 +95,14 @@ export default {
-      ]
+      ],
+      on: {
+        click: e => {
```
1 line this, `click: e => e.stopPropagation`

johnleider commented on src/components/VDialog/VDialog.js:
```diff
@@ -112,7 +132,10 @@ export default {
     children.push(h('div', {
-      'class': 'dialog__content',
+      'class': this.contentClasses,
+      style: {
```
1 line this

johnleider commented on src/components/VMenu/VMenu.js:
```diff
@@ -25,7 +26,8 @@ export default {
     Position,
-    Toggleable
+    Toggleable,
+    Dependent
```
Update to match this spec https://docs.google.com/document/d/1xWWnAw2zYce3sVvdiRC0OuIhNx1PCt-oCNzTXeREFaw/edit?usp=sharing

johnleider commented on src/components/VMenu/mixins/menu-generators.js:
```diff
@@ -40,13 +40,22 @@ export default {
+
       const options = {
-        'class': (`menu__content ${this.contentClass}`).trim(),
+        'class': [(`menu__content ${this.contentClass}`).trim(), { 'menuable__content__active': this.isActive }],
```
break this into 2 lines

johnleider commented on src/components/VSelect/VSelect.js:
```diff
@@ -45,7 +46,8 @@ export default {
     VMenu,
-    VBtn
+    VBtn,
+    Dependent
```
Update to match this spec https://docs.google.com/document/d/1xWWnAw2zYce3sVvdiRC0OuIhNx1PCt-oCNzTXeREFaw/edit?usp=sharing

johnleider commented on src/components/VTooltip/VTooltip.js:
```diff
@@ -5,11 +5,12 @@ import Colorable from '../../mixins/colorable'
   name: 'v-tooltip',
 
-  mixins: [Colorable, Detachable, Menuable, Toggleable],
+  mixins: [Colorable, Detachable, Menuable, Toggleable, Dependent],
```
Update to match this spec https://docs.google.com/document/d/1xWWnAw2zYce3sVvdiRC0OuIhNx1PCt-oCNzTXeREFaw/edit?usp=sharing

johnleider commented on src/directives/click-outside.js:
```diff
@@ -1,31 +1,44 @@
+  elements.push(el)
+
+  if (!clickedInEls(e, elements) && cb(e)) {
+    setTimeout(() => v.context.isActive = false, 0)
```
Not sure why lint didn't pick this up, but functions returning assignment should be wrapped in parenthesis

johnleider commented on src/directives/click-outside.js:
```diff
@@ -1,31 +1,44 @@
+  const b = el.getBoundingClientRect()
+  const x = e.clientX
+  const y = e.clientY
+  return x >= b.left && x <= b.right && y >= b.top && y <= b.bottom
```
Please add comments on what this is doing

johnleider commented on src/directives/click-outside.js:
```diff
@@ -1,31 +1,44 @@
   bind (el, binding, v) {
     v.context.$vuetify.load(() => {
-      const outside = document.querySelector('[data-app]')
       const click = e => directive(e, el, binding, v)
```
I understand it was like this before, but since you are in here and I'm trying to move things over to a common syntax, all event methods and variables that reference events should be pre-pended with **on**. So in this event, it would be **onClick** as the variable name.

johnleider commented on src/mixins/dependent.js:
```diff
@@ -0,0 +1,60 @@
+          const search = (children) => {
+            for (const child of children) {
+              if (child.isActive && (child.isDependent || (child.closeDependents && child.isDependent !== false))) {
+                results.push(child)
```
block nesting is too deep

johnleider commented on src/mixins/dependent.js:
```diff
@@ -0,0 +1,60 @@
+        for (const dependent of this.getOpenDependents()) {
+          result.push(...(dependent.getClicableDependentElements()))
+        }
+        return result
```
could this not be:

`
return this.getOpenDependents().map(d => d.getClickableDependentElements())
`

johnleider commented on src/mixins/dependent.js:
```diff
@@ -0,0 +1,60 @@
+        }
+        return result
+      },
+      getClicableDependentElements () {
```
method is mis-spelled

johnleider commented on src/mixins/dependent.js:
```diff
@@ -0,0 +1,60 @@
+      getClicableDependentElements () {
+        const result = [this.$el]
+        if (this.$refs.content) result.push(this.$refs.content)
+        result.push(...(this.getOpenDependentElements()))
```
is the extra parenthesis needed around **getOpen..**

johnleider commented on src/mixins/overlayable.js:
```diff
@@ -35,13 +35,18 @@ export default {
-      this.overlay.onclick = () => {
+      /*
+      // click-outide directive should handle this
+      this.overlay.onclick = (e) => {
```
you can remove this

Webifi commented:
Committed some code cleanup changes.

Many (all?) of the remaining codebeat issues seem to stem from wrapping a few of the mixins in a factory function, similar to some of your existing mixins.  Perhaps I shouldn't have followed that model?

Webifi commented:
Looks like I buggered something while cleaning up the code...  I'll look into it more tomorrow.

Webifi commented:
Bug I introduced should be fixed now. 

I believe the only remaining code formatting issues are to do with using a factory function for the new mixins.  Do these need to be refactored, or can they pass as is?

johnleider commented on src/components/VDialog/VDialog.js:
```diff
@@ -79,9 +101,16 @@ export default {
+          name: 'click-outside',
+          value: {
+            callback: this.closeConditional,
+            include: () => this.getOpenDependentElements()
```
change to `include: this.getOpenDependentElements`

johnleider commented on src/components/VMenu/mixins/menu-generators.js:
```diff
@@ -39,14 +39,28 @@ export default {
+      directives.push({
+        name: 'show',
+        value: this.isContentActive
+      })
```
space between return statement

johnleider commented on src/components/VMenu/mixins/menu-generators.js:
```diff
@@ -39,14 +39,28 @@ export default {
     genContent () {
       const options = {
-        'class': (`menu__content ${this.contentClass}`).trim(),
+        'class': [(`menu__content ${this.contentClass}`).trim(), { 'menuable__content__active': this.isActive }],
```
break this into new lines

johnleider commented on src/mixins/dependent.js:
```diff
@@ -0,0 +1,50 @@
+    } else {
+      results.push(...searchChildren(child.$children))
+    }
+  }
```
space between return statement

johnleider commented on src/mixins/dependent.js:
```diff
@@ -0,0 +1,50 @@
+      for (const dependent of this.getOpenDependents()) {
+        result.push(...dependent.getClickableDependentElements())
+      }
+      return result
```
space between return statement

johnleider commented on src/mixins/dependent.js:
```diff
@@ -0,0 +1,50 @@
+      const result = [this.$el]
+      if (this.$refs.content) result.push(this.$refs.content)
+      result.push(...this.getOpenDependentElements())
+      return result
```
space between return statement

johnleider commented on src/mixins/stackable.js:
```diff
@@ -0,0 +1,39 @@
+        // Return current zindex if not active
+        return getZIndex(content)
+      }
+      // Return max current z-index (excluding self) + 2 (2 to leave room for an overlay below, if needed)
```
space between return statement

johnleider commented on src/mixins/stackable.js:
```diff
@@ -0,0 +1,39 @@
+          zis.push(getZIndex(activeElement))
+        }
+      }
+      return Math.max(...zis)
```
space between return statement

johnleider commented on src/util/helpers.js:
```diff
@@ -115,3 +115,10 @@ export function getObjectValueByPath (obj, path) {
+  if (!el || el.nodeType !== Node.ELEMENT_NODE) return 0
+  var zi = document.defaultView.getComputedStyle(el).getPropertyValue('z-index')
+  if (isNaN(zi)) return getZIndex(el.parentNode)
+  return zi
```
space between return statement

johnleider commented on src/mixins/dependent.js:
```diff
@@ -0,0 +1,50 @@
+
+  watch: {
+    isActive (val) {
+      if (!val) {
```
to reduce assignment here,

`if (val) return`

johnleider commented on src/mixins/dependent.js:
```diff
@@ -0,0 +1,50 @@
+  methods: {
+    getOpenDependents () {
+      if (this.closeDependents) return searchChildren(this.$children)
+      return []
```
space between return statement

johnleider commented on src/directives/click-outside.js:
```diff
@@ -1,31 +1,65 @@
+  // (we're in capturing event and we want to check for multiple elements,
+  //  so can't use target.)
+  const b = el.getBoundingClientRect()
+  // Check if the click was in the element's bounding rect
```
space between return statement

johnleider commented on src/directives/click-outside.js:
```diff
@@ -1,31 +1,65 @@
+  for (const el of elements) {
+    if (clickedInEl(el, x, y)) return true
+  }
+  return false
```
space between return statement

johnleider commented:
This should be my last round of updates, thank you.

Webifi commented:
Happy to help.  More a selfish endeavor, since I need the functionality.

I hope there aren't a ton of unforeseen issues this will cause, but I'm pretty confident it'll fix more than it creates.   And it should lay some ground work to potentially fix a few more issues later.
